math/src/main/kotlin/ktx/math/ImmutableVector.kt

```diff
@@ -0,0 +1,154 @@
+@file:Suppress("NOTHING_TO_INLINE")
+
+package ktx.math
+
+import com.badlogic.gdx.math.Interpolation
+import com.badlogic.gdx.math.MathUtils
+import com.badlogic.gdx.math.Vector
+import kotlin.math.abs
+import kotlin.math.sqrt
+
+/**
+ * Represent an immutable vector.
+ */
+interface ImmutableVector<T : ImmutableVector<T>> : Comparable<T> {
+
+    /**
+     * Returns the squared euclidean length
+     *
+     * This method is faster than [Vector.len] because it avoids calculating a square root. It is useful for comparisons,
+     * but not for getting exact lengths, as the return value is the square of the actual length.
+     */
+    val len2: Float
+
+    /** Returns whether the length of this vector is smaller than the given [margin] */
+    fun isZero(margin: Float = MathUtils.FLOAT_ROUNDING_ERROR): Boolean
+
+    /** Returns the result of subtracting the [other] vector from this vector */
+    operator fun minus(other: T): T
+
+    /** Returns the result of adding the [other] vector to this vector */
+    operator fun plus(other: T): T
+
+    /** Returns the same vector with all members incremented by 1 */
+    operator fun inc(): T
+
+    /** Returns the same vector with all members decremented by 1 */
+    operator fun dec(): T
+
+    /** Returns this vector scaled by the given [scalar] */
+    operator fun times(scalar: Float): T
+
+    /** Returns his vector scaled by the given [vector] */
+    operator fun times(vector: T): T
+
+    /** Returns he dot product of this vector by the given [vector] */
+    infix fun dot(vector: T): Float
+
+    /**
+     * Returns the squared distance between this and the other vector
+     *
+     * This method is faster than [dst] because it avoids calculating a square root. It is useful for comparisons,
+     * but not for getting exact distance, as the return value is the square of the actual distance.
+     */
+    infix fun dst2(vector: T): Float
+
+    /** Linearly interpolates between this vector and the target vector by alpha */
+    fun lerp(target: T, alpha: Float): T
+
+    /** Returns a vector of same length and a random direction */
+    fun withRandomDirection(): T
+
+    /** Returns true if this vector is in line with the other vector (either in the same or the opposite direction) */
+    fun isOnLine(other: T, epsilon: Float = MathUtils.FLOAT_ROUNDING_ERROR): Boolean
+
+    /** Compares this vector with the other vector, using the supplied epsilon for fuzzy equality testing */
+    fun epsilonEquals(other: T, epsilon: Float): Boolean
+
+    override fun compareTo(other: T): Int = len2.compareTo(other.len2)
+}
+
+/**
+ * Returns the euclidean length
+ */
+inline val ImmutableVector<*>.len: Float get() = sqrt(len2)
+
+/**
+ * Returns the unit vector of same direction or this vector if it is zero.
+ */
+val <T : ImmutableVector<T>> T.nor: T
+    get() {
+        val l2 = len2
```

Comment:
Can you rename the `l2` parameter to `lengthSquared` for readability? I want the sources to be easy to read for Kotlin/LibGDX beginners.

Reply:
I changed for the folowing:

```kotlin
override val nor: ImmutableVector2 get() = withLength2(1f)

override fun withLength2(length2: Float): ImmutableVector2 {
  val oldLen2 = this.len2

  return if (oldLen2 == 0f || oldLen2 == length2) this else times(sqrt(length2 / oldLen2))
}
```

Is that ok?